wordpress-mobile/WordPress-iOS
# Add A/B test for site intent question

Fixes https://github.com/wordpress-mobile/WordPress-iOS/issues/17988

The A/B implementation proposed here is taken from [a previous implementation](https://github.com/wordpress-mobile/WordPress-iOS/pull/17328/files#diff-dd5d1197f8b51946ef1a85fd4ab79950224beb9abf9b4b20862683671230428d) which was used for an A/B test and then removed when the test concluded. The hash of the user's authorization token to assign the user to one of two groups: Control or Treatment. Users in the Treatment group will see the Site Intent Question screen, users in Control will not.

If we decide to ship this screen to all users after the A/B test finishes, `SiteIntentAB` and related code can be removed from the codebase.

At time of writing, the A/B test is configured to roll the site intent question screen out to 50% of WP.com users.

For now, the site intent screen is blank since it's out-of-scope of this PR. https://github.com/wordpress-mobile/WordPress-iOS/issues/18106 and other issues track the implementation of the actual screen contents.

## To test

The prerequisites to test this PR are:
- Turn ON the feature flag added in https://github.com/wordpress-mobile/WordPress-iOS/pull/18063 (found in the app under Me > App Settings > Debug). I don't know of a way to turn the flag on before logging in, so the workaround is to log in using any user, turn flag ON, and log in with the account you want to test. The flag's state will be persisted.
- Use a WP.com account that falls into the Treatment group (details on this below). There's no way to identify this upfront, so instead you must keep trying different accounts until the new screen (currently blank as of time of writing) appears at the start of the site creation process. You can then use that account for testing, knowing that it is in the Treatment group. Any account that is not in the treatment group is said to be in the Control group. Reinstalling the app may change the token, thus changing the user's group.

The following sections cover the test cases for this PR.

```
#### Test feature 
- [ ] Create new account and new site (via "Create a new site" button on the "Welcome to WordPress" screen)
- [ ] Create new site using account that has no sites (via "Add new site" on the "My Site" screen)
- [ ] Create new site using account that has existing sites (via "Create a new site" button on the post-login site list screen)
- [ ] Create new site using account that has existing sites (via plus button on site switcher, "My Sites", screen)

#### Check for regressions
- [ ] Verify that the site intent question screen is not shown when the feature flag is OFF
- [ ] Verify that the site intent question screen is not shown for accounts in the Control group
- [ ] Verify that adding a new self-hosted site doesn't show the site intent question screen
```

## Regression Notes
1. Potential unintended areas of impact

This change conditionally adds a new step to the site creation flow for users in the Treatment group of the A/B test. We verify that the Control group is unaffected by this change. Also, while this screen is developed, it's behind a feature flag that is OFF by default. We must ensure it stays off until we ship this feature.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

- I verified that when the feature flag is OFF, the new screen is not shown, even when the user is in the Treatment group
- I verified that when the feature flag is ON, the new screen is not shown, if a user in the Control group 

3. What automated tests I added (or what prevented me from doing so)

I added unit tests, `SiteCreationIntentTests`, to verify that the site intent question is shown under the right conditions (feature flag, A/B test, etc).

PR submission checklist:
